### DISCUSSION TITLE
Implementing Ex.36 from scikit-fem in FElupe

### DISCUSSION THREAD

**Discussion by adtzlr** [adtzlr/felupe#615] [Show and tell]
Here are some possible implementations of [Ex.36](https://github.com/kinnala/scikit-fem/blob/master/docs/examples/ex36.py), taken from the docs of `scikit-fem`. These examples require `pip install felupe[all] matadi` to be installed.

```python
import felupe as fem

mesh = fem.Cube(n=5).triangulate().add_midpoints_edges()
region = fem.RegionQuadraticTetra(mesh)
field = fem.FieldsMixed(region, n=2)
```

The mixed-field container automatically creates the Taylor-Hood element. The region of the first field is the one we created (`RegionQuadraticTetra`)

```python
field[0].region
```

```
<felupe Region object>
  Element formulation: QuadraticTetra
  Quadrature rule: Tetrahedron
  Gradient evaluated: True
```

and the second field is based on a region one order lower, [`RegionTetra`](https://felupe.readthedocs.io/en/latest/felupe/region.html#felupe.RegionTetra) - but with the quadrature scheme from `RegionQuadraticTetra`.

```python
field[1].region
```

```
<felupe Region object>
  Element formulation: Tetra
  Quadrature rule: Tetrahedron
  Gradient evaluated: False
```

Boundary conditions are created with a template for uniaxial tension. For more details on boundary conditions please see [`Boundary`](https://felupe.readthedocs.io/en/latest/felupe/dof.html#felupe.Boundary).

```python
boundaries, loadcase = fem.dof.uniaxial(field, clamped=False)
```

In this first implementation, we'll use the gradients of the strain energy function, carried out by automatic differentiation using [matadi](https://github.com/adtzlr/matadi).

```python
import matadi as mat
from matadi.math import det, gradient, log, sqrt, trace

F = mat.Variable("F", 3, 3)
p = mat.Variable("p", 1, 1)
ζ = mat.Variable("ζ", 1, 1)


def ψ(x, μ, λ):
    F, p, ζ = x[0], x[1], x[2]

    J = det(F)
    Js = (λ + p + sqrt((λ + p) ** 2 + 4 * λ * μ)) / (2 * λ)

    C = F.T @ F
    I1 = trace(C)

    ψs = p * (J - Js) + μ / 2 * (I1 - 3) - μ * log(Js) + λ / 2 * (Js - 1) ** 2
    ζ_new = ζ

    return gradient(ψs, F), gradient(ψs, p), ζ_new


umat = mat.MaterialTensor(x=[F, p, ζ], fun=ψ, kwargs=dict(μ=1, λ=1e4), statevars=1)
solid = fem.SolidBody(umat, field)

move = fem.math.linsteps([0, 1], num=5)
step = fem.Step(items=[solid], ramp={boundaries["move"]: move}, boundaries=boundaries)

job = fem.CharacteristicCurve(steps=[step], boundary=boundaries["move"])
job.evaluate()
fig, ax = job.plot(
    xlabel="Displacement $u$ in mm $\longrightarrow$",
    ylabel="Normal Force $F$ in N $\longrightarrow$",
)

ax2 = field.imshow("Principal Values of Logarithmic Strain")
```

![image](https://github.com/adtzlr/felupe/assets/5793153/240c7838-5dc0-4bf6-96cf-34fc6f43656c)

![image](https://github.com/adtzlr/felupe/assets/5793153/6cc1ca82-e7ca-4bbd-83fe-5b4bc0fc55a6)

**Comment by adtzlr** [comment]
FElupe also has a very similar built-in [`ThreeFieldVariation`](https://felupe.readthedocs.io/en/latest/felupe/constitution.html#felupe.ThreeFieldVariation) (u,p,J). It basically slices out the volumetric part of your strain energy function, adds the constraints between (u,p,J) and eliminates volumetric locking. This time we'll use the [`Hyperelastic`](https://felupe.readthedocs.io/en/latest/felupe/constitution.html#felupe.Hyperelastic) material definition from FElupe. Once again with automatic differentiation but with [tensortrax](https://github.com/adtzlr/tensortrax) as backend.

Here's the code:

```python
import felupe as fem

mesh = fem.Cube(n=5).triangulate().add_midpoints_edges()
region = fem.RegionQuadraticTetra(mesh)
field = fem.FieldsMixed(region, n=3)

boundaries, loadcase = fem.dof.uniaxial(field, clamped=False)

import tensortrax.math as tm

def ψ(C, μ, λ):
    
    I1 = tm.trace(C)
    J = tm.sqrt(tm.linalg.det(C))
    lnJ = tm.log(J)
    
    return μ / 2 * (I1 - 3) - μ * lnJ + λ / 2 * (J - 1) ** 2

umat = fem.ThreeFieldVariation(fem.Hyperelastic(ψ, μ=1, λ=1e4, parallel=True))
solid = fem.SolidBody(umat, field)

move = fem.math.linsteps([0, 1], num=5)
step = fem.Step(items=[solid], ramp={boundaries["move"]: move}, boundaries=boundaries)

job = fem.CharacteristicCurve(steps=[step], boundary=boundaries["move"])
job.evaluate(tol=1e-2)
fig, ax = job.plot(
    xlabel="Displacement $u$ in mm $\longrightarrow$",
    ylabel="Normal Force $F$ in N $\longrightarrow$",
)

ax2 = field.imshow("Principal Values of Logarithmic Strain")
```

**Reply by adtzlr** [reply]
FYI, there is already a code-block for **finite-strain-viscoelasticity**, please see the docstring of [`Hyperelastic`](https://felupe.readthedocs.io/en/latest/felupe/constitution.html#felupe.Hyperelastic).

**Comment by adtzlr** [comment]
The strain energy function from Ex.36 is implemented in FElupe as [`NeoHookeCompressible`](https://felupe.readthedocs.io/en/latest/felupe/constitution.html#felupe.NeoHookeCompressible). Instead of defining a custom strain energy function, one could also use this class instead.

```python
import felupe as fem

mesh = fem.Cube(n=5).triangulate().add_midpoints_edges()
region = fem.RegionQuadraticTetra(mesh)
field = fem.FieldsMixed(region, n=3)

boundaries, loadcase = fem.dof.uniaxial(field, clamped=False)

umat = fem.ThreeFieldVariation(fem.NeoHookeCompressible(mu=1, lmbda=1e4))
solid = fem.SolidBody(umat, field)

move = fem.math.linsteps([0, 1], num=5)
step = fem.Step(items=[solid], ramp={boundaries["move"]: move}, boundaries=boundaries)

job = fem.CharacteristicCurve(steps=[step], boundary=boundaries["move"])
job.evaluate(tol=1e-2)
fig, ax = job.plot(
    xlabel="Displacement $u$ in mm $\longrightarrow$",
    ylabel="Normal Force $F$ in N $\longrightarrow$",
)

ax2 = field.imshow("Principal Values of Logarithmic Strain")
```

**Comment by adtzlr** [comment]
a) For the most general approach and without all the fancy automatic differentiation stuff, please have a look at [`Material`](https://felupe.readthedocs.io/en/latest/felupe/constitution.html#felupe.Material). Unfortunately, I only have a minimal documentation docstring ready for mixed-fields yet.

b) For a more scikit-fem approach, there is also [`Form`](https://felupe.readthedocs.io/en/latest/felupe/assembly.html#felupe.Form) along with [`FormItem`](https://felupe.readthedocs.io/en/latest/felupe/assembly.html#felupe.FormItem).

This should be **really** enough to read for the beginning 😄 .

**Comment by bhaveshshrimali** [comment]
Great @adtzlr ! This is super helpful!! Many thanks

**Comment by adtzlr** [comment]
And here you can find more or less the 1:1 translation of the original code. I replaced only the math-helpers. I used `pypardiso` as a faster alternative to the scipy-sparse solver (`pip install pypardiso`). Instead of `Material`,  it is also possible to create the class by hand. There is nothing special except for the shape of the state variables, see [Constitution](https://felupe.readthedocs.io/en/latest/felupe/constitution.html).

```python
import numpy as np
import felupe as fem
from felupe.math import det, inv, transpose, identity
import pypardiso


def F1(F, p, mu, lmbda):
    J = det(F)
    Finv = inv(F)
    return p * J * transpose(Finv) + mu * F


def F2(F, p, mu, lmbda):
    J = det(F)
    Js = 0.5 * (lmbda + p + 2.0 * np.sqrt(lmbda * mu + 0.25 * (lmbda + p) ** 2)) / lmbda
    dJsdp = (
        0.25 * lmbda + 0.25 * p + 0.5 * np.sqrt(lmbda * mu + 0.25 * (lmbda + p) ** 2)
    ) / (lmbda * np.sqrt(lmbda * mu + 0.25 * (lmbda + p) ** 2))
    return J - (Js + (p + mu / Js - lmbda * (Js - 1)) * dJsdp)


def A11(F, p, mu, lmbda):
    eye = identity(F)
    J = det(F)
    Finv = inv(F)
    L = (
        p * J * np.einsum("lk...,ji...->ijkl...", Finv, Finv)
        - p * J * np.einsum("jk...,li...->ijkl...", Finv, Finv)
        + mu * np.einsum("ik...,jl...->ijkl...", eye, eye)
    )
    return L


def A12(F, p, mu, lmbda):
    J = det(F)
    Finv = inv(F)
    return J * transpose(Finv)


def A22(F, p, mu, lmbda):
    Js = 0.5 * (lmbda + p + 2.0 * np.sqrt(lmbda * mu + 0.25 * (lmbda + p) ** 2)) / lmbda
    dJsdp = (
        0.25 * lmbda + 0.25 * p + 0.5 * np.sqrt(lmbda * mu + 0.25 * (lmbda + p) ** 2)
    ) / (lmbda * np.sqrt(lmbda * mu + 0.25 * (lmbda + p) ** 2))
    d2Jdp2 = 0.25 * mu / (lmbda * mu + 0.25 * (lmbda + p) ** 2) ** (3 / 2)
    L = (
        -2.0 * dJsdp
        - p * d2Jdp2
        + mu / Js**2 * dJsdp**2
        - mu / Js * d2Jdp2
        + lmbda * (Js - 1.0) * d2Jdp2
        + lmbda * dJsdp**2
    )
    return L


def stress(x, mu, lmbda):
    F, p, statevars = x[0], x[1], x[-1]
    return [F1(F, p, mu, lmbda), F2(F, p, mu, lmbda), statevars]


def elasticity(x, mu, lmbda):
    F, p = x[0], x[1]
    return [A11(F, p, mu, lmbda), A12(F, p, mu, lmbda), A22(F, p, mu, lmbda)]


mesh = fem.Cube(n=5).triangulate().add_midpoints_edges()

region_u = fem.RegionQuadraticTetra(mesh=mesh)
region_p = fem.RegionTetra(mesh=fem.mesh.dual(mesh, disconnect=False), quadrature=region_u.quadrature, grad=False)

displacement = fem.Field(region_u, dim=3)
pressure = fem.Field(region_p)
field = fem.FieldContainer([displacement, pressure])

boundaries, loadcase = fem.dof.uniaxial(field, clamped=False)
umat = fem.Material(stress, elasticity, mu=1, lmbda=1e4)
solid = fem.SolidBody(umat, field)

move = fem.math.linsteps([0, 1], num=5)
step = fem.Step(items=[solid], ramp={boundaries["move"]: move}, boundaries=boundaries)

job = fem.CharacteristicCurve(steps=[step], boundary=boundaries["move"])
job.evaluate(tol=1e-2, solver=pypardiso.spsolve)
fig, ax = job.plot(
    xlabel="Displacement $u$ in mm $\longrightarrow$",
    ylabel="Normal Force $F$ in N $\longrightarrow$",
)

ax2 = field.imshow("Principal Values of Logarithmic Strain")
```

**Comment by adtzlr** [comment]
`FieldsMixed(region, n=2)` for `RegionQuadraticTetra` is just a template for the following lines of code:

```python
mesh = fem.Cube(n=5).triangulate().add_midpoints_edges()

region_u = fem.RegionQuadraticTetra(mesh=mesh)
region_p = fem.RegionTetra(mesh=fem.mesh.dual(mesh, disconnect=False), quadrature=region_u.quadrature, grad=False)

displacement = fem.Field(region_u, dim=3)
pressure = fem.Field(region_p)
field = fem.FieldContainer([displacement, pressure])
```

**Note**: The *dual* mesh for the secondary field $p$ may be `disconnect`-ed or not, this is your choice. It seems to be more stable if it is not disconnected here.

A german [Wikipedia article](https://de.wikipedia.org/wiki/Gemischte_finite_Elemente) contains a note on the instability when dealing with a disconnected mesh/field for the pressure. @bhaveshshrimali Do you know more on that topic?

**Reply by bhaveshshrimali** [reply]
@adtzlr the above is the Taylor-Hood mixed-element, whereas the wikipedia article suggests that a P2-P0 approximation (P0 being the discontinuous pressure field) could work too by imposing a stabilization parameter (the approximation, by itself, isn't inf-sup stable). I haven''t taken a look at the reference book for that discussion. 

Another alternative is to use: P2+bubble in 2D and P2+facet-bubble+internal-bubble in 3D along with P1-discontinuous for pressure (similar to disconnected pressure that you wrote above; I will take a closer look at the Region objects). These were the ones I added to scikit-fem back in the day for some of the work I was doing (but these are very heavy in  terms of the number of degrees of freedom per elem). 

The basis functions are here: http://pamies.cee.illinois.edu/Publications_files/JMPS2017b.pdf  or https://github.com/kinnala/scikit-fem/blob/master/skfem/element/element_tet/element_tet_ccr.py (for the displacement). The pressure should be simple P1-discontinuous.

https://link.springer.com/book/10.1007/978-3-642-87047-7
http://archive.numdam.org/article/M2AN_1973__7_3_33_0.pdf

**Reply by adtzlr** [reply]
Thanks for the explanations and references! I meant the comments in the Wiki article on P2-P1(disc.). Nevermind, it's a german article on Wikipedia - no actual scientific reference. Personally, I found the Q1-Q0 formulations to be the best choice in industry-related simulations for nearly-incompressible hyperelastic problems. But there are controversal discussions in scientific articles/books on that element formulation... :smile:

**Reply by bhaveshshrimali** [reply]
Oh, I have never seen a P2-P1(disc) being used anywhere. That it is unstable for general meshes is not surprising. I did check out Ganesan and Tobiska 8.6 which essentially refers to the same element I linked above, but I couldn't find a discussion on P2-P1(disc) beyond what is already provided in the article (I read the english translation 😄  ). P1(disc) is useful for mass conservation in Stokes, and for heterogeneous materials (for capturing stress discontinuities more accurately) in elasticity. 

![image](https://github.com/adtzlr/felupe/assets/22624037/8287d477-5d26-4c02-95e3-c192fa28230f)


![image](https://github.com/adtzlr/felupe/assets/22624037/8404fa2a-181f-497f-ae61-c83bdb3ab068)

Q1-Q0 does not satisfy inf-sup, right ?

**Reply by bhaveshshrimali** [reply]
As a side note, the assembly is remarkably fast in felupe. I will open separate discussions as I grok through the code, but running example 36 for a similar mesh was super-fast (even with no numba etc.). Great work on this. Thanks!

**Reply by adtzlr** [reply]
> As a side note, the assembly is remarkably fast in felupe. I will open separate discussions as I grok through the code, but running example 36 for a similar mesh was super-fast (even with no numba etc.). Great work on this. Thanks!

Sounds great. But please compare mesh points /dof, PyVista shows all internal edges too and the mesh could look finer than it is!

See https://github.com/pyvista/pyvista/issues/867#issue-678805909

**Reply by bhaveshshrimali** [reply]
I used `mesh.ncells` as a metric to compare number of cells, and visualized in Paraview.

**Reply by adtzlr** [reply]
That are indeed good news 👍🏻. I'm quite happy with assembly runtimes now. However, the overall memory footprint is still quite massive and lots of substeps take their time to complete because of array (re-)creations / allocations. I try to use the `out`-argument of numpy-functions when possible but there is still a lot to improve.

**Reply by adtzlr** [reply]
> Q1-Q0 does not satisfy inf-sup, right ?

Yes, at least from what I read and understand :smile: on that topic, you're right.
